lupa/_lupa.pyx

```diff
@@ -1556,16 +1640,26 @@ cdef bint call_python(LuaRuntime runtime, lua_State *L, py_object* py_obj) excep
     # Callers must assure that py_obj.obj is not NULL, i.e. it points to a valid Python object.
     cdef int i, nargs = lua.lua_gettop(L) - 1
     cdef tuple args
+    cdef dict kwargs
 
     f = <object>py_obj.obj
 
-    if not nargs:
+    if nargs == 0:
         lua.lua_settop(L, 0)  # FIXME
         result = f()
     else:
-        arg = py_from_lua(runtime, L, 2)
+        args = ()
+        kwargs = {}
+        
+        for i in range(nargs):
+            arg = py_from_lua(runtime, L, i+2)
+            if isinstance(arg, _PyArguments):
+                args += (<_PyArguments>arg).args
+                kwargs = dict(**kwargs, **(<_PyArguments>arg).kwargs)
+            else:
+                args += (arg, )
```

Comment:
Hmm. Now that I see the `args += (arg,)` – most of the time, there won't be any `_PyArguments` on the stack. Building an args tuple here one by one seems rather wasteful in that regard.

I think you were right with your original approach to build a list instead, and just use `.append()` and `.extend()`.

Alternatively, there is a documented [`_PyTuple_Resize()`](https://docs.python.org/3/c-api/tuple.html#c._PyTuple_Resize) function which would fit this use case, while allowing us to allocate a pre-sized tuple. It's also supported by PyPy, I think.

Reply:
I like the second solution best. It sounds more efficient to allocate the whole tuple beforehand and then shrink it if necessary. I'll look into it.